libevent-sys/build.rs

```diff
@@ -1,18 +1,163 @@
+            .define("EVENT__DISABLE_SAMPLES", "ON");
+
+        // TODO: Address policy warnings for cross-compilation?
+        /*config.define("CMAKE_POLICY_DEFAULT_CMP0056", "NEW")
+            .define("CMAKE_POLICY_DEFAULT_CMP0066", "NEW")*/
+    }
+
+    // TODO: Can we tap into "-vv" cargo argument?
+    if cfg!(feature = "verbose_build") {
+        config.very_verbose(true);
+    }
+
+    let dst = config.build();
+
+    println!("cargo:rustc-link-search={}/lib", dst.display());
+
+    // Library 'event' is considered deprecated, so link each sub-component
+    // individually.
+    println!("cargo:rustc-link-lib=static=event_core");
+    println!("cargo:rustc-link-lib=static=event_extra");
+
+    if cfg!(feature = "openssl") || cfg!(feature = "openssl_bundled") {
+        println!("cargo:rustc-link-lib=static=event_openssl");
+    }
+
+    if cfg!(feature = "threading") {
+        println!("cargo:rustc-link-lib=static=event_pthreads");
+    }
+
+    println!("cargo:include={}/include", dst.display());
+
+    dst
+}
+
+#[cfg(not(feature = "bundled"))]
+fn run_pkg_config() -> Option<Vec<String>> {
+    let mut pkg = pkg_config::Config::new();
+    pkg.cargo_metadata(false)
         .atleast_version("2")
-        .statik(cfg!(feature = "static"))
-        .probe("libevent")
-        .unwrap();
+        .statik(cfg!(feature = "static"));
+
+    let mut lib = match pkg.probe("libevent_core") {
+        Ok(lib) => lib,
+        Err(_e) => { return None; }
+    };
+    // TODO: Probably combine all pc includes, just to be safe.
+
+    pkg.cargo_metadata(true).probe("libevent_core").unwrap();
+    pkg.cargo_metadata(true).probe("libevent_extra").unwrap();
+
+    if cfg!(feature = "openssl") {
+        pkg.cargo_metadata(true).probe("libevent_openssl").unwrap();
+    }
+
+    if cfg!(feature = "threading") {
+        pkg.cargo_metadata(true).probe("libevent_pthreads").unwrap();
+    }
+
+    let include_paths = lib.include_paths.drain(..).map(|path| {
+        let path_s = path.to_str().unwrap().to_string();
+        println!("cargo:include={}", &path_s);
+        path_s
+    }).collect();
+
+    Some(include_paths)
+}
+
+#[cfg(feature = "bundled")]
+fn find_libevent() -> Option<Vec<String>> {
+    use std::path::Path;
+    use std::process::Command;
+
+    if !Path::new("libevent/.git").exists() {
+        Command::new("git").args(&["submodule", "update", "--init"])
+            .status().expect("Running `git submodule init` failed.");
+    } else {
+        Command::new("git").args(&["submodule", "update", "--recursive"])
+            .status().expect("Running `git submodule update` failed.");
+    }
+
+    Some(vec![format!("{}/include", build_libevent("libevent").display())])
+}
+#[cfg(not(feature = "bundled"))]
+fn find_libevent() -> Option<Vec<String>> {
+    run_pkg_config()
+}
+
+fn main() {
+    println!("cargo:rerun-if-changed=build.rs");
+    println!("cargo:rerun-if-changed=libevent");
```

Comment:
If you care about more than the last mtime of the `libevent` directory then this may not cut it as is.

> If the path points to a directory, it does not automatically traverse the directory for changes. Only the mtime change of the directory itself is considered (which corresponds to some types of changes within the directory, depending on platform). To request a re-run on any changes within an entire directory, print a line for the directory and separate lines for everything inside it, recursively.

Reply:
Works well enough for this case I think (covers source file changes, as well as updating the submodule rev). And I believe all this is generally done in `CARGO_HOME` anyway, so if you're messing around in there..

Reply:
Ah sure that makes sense.